File: medium/198. House Robber/198. House Robber.ts
function rob(nums: number[]): number {
    if (!nums.length)
        return 0
    if (nums.length === 1)
        return nums[0]

    let prev1 = 0
    let prev2 = nums[0]

    for (let i = 1; i < nums.length; i++) {
        const max_rob = Math.max(prev2, prev1 + nums[i])
        prev1 = prev2
        prev2 = max_rob
    }

    return prev2
};

File: medium/198. House Robber/198. House Robber.go
package main

import "slices"

func rob(nums []int) int {
	n := len(nums)
	// prev1 := 0
	// prev2 := nums[0]
	dp := slices.Repeat([]int{0}, n)
	dp[0] = nums[0]
	dp[1] = max(nums[0], nums[1])

	for i := 2; i < n; i++ {
		// max_rob := max(prev2, prev1+nums[i])
		// prev1 = prev2
		// prev2 = max_rob
		dp[i] = max(dp[i-1], dp[i-2]+nums[i])
	}

	// return prev2
	return dp[n-1]
}
